### ISSUE TITLE
Scan : Take screenshots of failed UI Tests 

### ISSUE THREAD

**Issue by FabienLydoire** [fastlane/fastlane#12316]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
The issue #1932 "App screenshots of failed steps" has been closed without actual implementation.
It would be nice to get the screenshots of failed UI tests using scan.
Is it possible with the current version ?

**Comment by Speakus** [comment]
this issue is related I guess: https://github.com/fastlane/fastlane/issues/6619

**Comment by Speakus** [comment]
this one too: https://github.com/fastlane/fastlane/issues/9358

**Comment by joshdholtz** [comment]
@FabienLydoire This isn't a feature but does sound like it would be pretty nice. We have a pretty nice backlog of things to work on so if you would feel comfortable prototyping this out and creating a PR that would be 💯 Otherwise we can add this feature to our list of things to work on when we can fit it in 🙃

**Comment by Speakus** [comment]
I did this fix: https://github.com/Speakus/fastlane/commit/65f2bcaa4b851926cc8bd7cb0f29f8069d2c6e00 but it is unfinished yet.

**Comment by Speakus** [comment]
unfortunately `--screenshots` flag does not work properly yet for `xcpretty`

**Comment by Speakus** [comment]
unfortunately `--screenshots` flag does not work properly yet for `xcpretty` https://github.com/supermarin/xcpretty/issues/251

**Comment by FabienLydoire** [comment]
I did manage to get a decent report with screenshots using "result_bundle" in scan parameters and passing the bundle to TitouanVanBelle/XCTestHTMLReport
The automatic screenshots taken by Xcode are displayed in the HTML report.

**Comment by Speakus** [comment]
@FabienLydoire could you share the link to solution?

**Comment by krin-san** [comment]
Xcode makes screenshots automatically but in `.jpg` format. Current implementation of xcpretty [filters out non-png file](https://github.com/supermarin/xcpretty/blob/7a072a1ec930ae3ab1b9ef4b8df5bc9a69ba0db0/lib/xcpretty/reporters/html.rb#L71). So right now I'm searching for a way to inject custom shell script between scan and xcpretty to convert jpg files to png.

@FabienLydoire, how do you managed to make it working? Please, don't just close the issue without sharing your experience.

**Comment by FabienLydoire** [comment]
Seen in this article [Mobile Automation Testing: Running iOS tests in parallel](https://medium.com/@akifev/mobile-automation-testing-running-ios-tests-in-parallel-d68bcb574fc4).

Install [XCTestHTMLReport](https://github.com/TitouanVanBelle/XCTestHTMLReport).
Update Fastfile to generate a report using xchtmlreport.

```
def generate_report
	puts "Generating Test Report ..."
	sh 'xchtmlreport -r test_output/MyXcodeProject.test_result'
	puts "Test Report Successfully generated"
end

 desc "Runs all the unit tests"
  lane :unit_test do |options|
	begin
		unit_test_private options
	rescue
		generate_report
	end
  end

  private_lane :unit_test_private do |options|
    setup_jenkins(result_bundle: false)
    scan(
		project: './MyXcodeProject.xcodeproj',
		device: 'iPhone 6 Plus',
		scheme: options[:scheme],
		clean: true,
		derived_data_path: './fastlane',
		output_directory: "./fastlane/test_output",
		result_bundle: "TestResults" # To generate test reports
    )
	generate_report
  end
```

In your ```./fastlane/test_output/MyXcodeProject.test_result``` you will have an ```index.html``` with your tests and screenshots.

**Comment by ybeapps** [comment]
## For those who copy @FabienLydoire's script.. 

When coping the following:
```
def generate_report
	puts "Generating Test Report ..."
	sh 'xchtmlreport -r test_output/MyXcodeProject.test_result'
	puts "Test Report Successfully generated"
end
```

Don't forget to change **MyXcodeProject** with your scheme name

**Comment by krin-san** [comment]
I'd like to share another idea of improvement here. Sometimes, when `scan` fails with compile exception it won't leave any files `xchtmlreport` could so it will throw it's own exception. If you want the `scan` exception to prevail in logs, wrap `xchtmlreport` execution and re-throw original exception like that:
```ruby
  lane :unit_test do |options|
    begin
      unit_test_private options
    rescue
      begin
        generate_report
      rescue
        # Ignore any exceptions coming from xchtmlreport
      end

      raise  # Re-throw original fastlane-scan exception
    end
  end
```

**Comment by FabienLydoire** [comment]
@ybeapps : thanks, I didn't catch it since my scheme and project have the same name
@krin-san : thanks for the proposal, great improvement !
